mlx/dtype_utils.h

```diff
@@ -0,0 +1,215 @@
+#define MLX_FORALL_INT_TYPES(_) \
+  _(uint8_t, uint8)             \
+  _(uint16_t, uint16)           \
+  _(uint32_t, uint32)           \
+  _(uint64_t, uint64)           \
+  _(int8_t, int8)               \
+  _(int16_t, int16)             \
+  _(int32_t, int32)             \
+  _(int64_t, int64)
+
+#define MLX_FORALL_FLOAT_TYPES(_) \
+  _(float16_t, float16)           \
+  _(float, float32)               \
+  _(double, float64)              \
+  _(bfloat16_t, bfloat16)
+
+// Calls the provided macro on every Dtype, providing the C type and the
+// Dtype name to each call.
+//
+// @param _ A macro that takes two parameters: the name of a C type, and the
+//          name of the corresponding Dtype enumerator.
+#define MLX_FORALL_DTYPES(_) \
+  MLX_FORALL_INT_TYPES(_)    \
+  MLX_FORALL_FLOAT_TYPES(_)  \
+  _(bool, bool_)             \
+  _(complex64_t, complex64)
+
+// Maps Dtypes to C++ types.
+template <Dtype::Val N>
+struct DtypeToCppType;
+
+#define SPECIALIZE_DtypeToCppType(CPP_TYPE, DTYPE) \
+  template <>                                      \
+  struct DtypeToCppType<Dtype::Val::DTYPE> {       \
+    using type = CPP_TYPE;                         \
+  };
+
+MLX_FORALL_DTYPES(SPECIALIZE_DtypeToCppType)
+
+#undef SPECIALIZE_DtypeToCppType
+
+// Maps C++ types to Dtypes.
+template <typename T>
+struct CppTypeToDtype;
+
+#define SPECIALIZE_CppTypeToDtype(CPP_TYPE, DTYPE) \
+  template <>                                      \
+  struct CppTypeToDtype<CPP_TYPE>                  \
+      : std::integral_constant<Dtype::Val, Dtype::Val::DTYPE> {};
+
+MLX_FORALL_DTYPES(SPECIALIZE_CppTypeToDtype)
+
+#undef SPECIALIZE_CppTypeToDtype
+
+// Helper macros for switch case macros (see below)
+//
+// These macros are not meant to be used directly. They provide an easy way to
+// generate a switch statement that can handle subsets of Dtypes supported.
+
+#define MLX_INTERNAL_SWITCH_CASE(enum_type, CTYPE_ALIAS, ...)         \
+  case enum_type: {                                                   \
+    using CTYPE_ALIAS = ::mlx::core::DtypeToCppType<enum_type>::type; \
+    __VA_ARGS__;                                                      \
+    break;                                                            \
+  }
+
+#define MLX_INTERNAL_SWITCH(TYPE, ...) \
+  [&] {                                \
+    switch (TYPE) { __VA_ARGS__ }      \
+  }()
+
+#define MLX_INTERNAL_SWITCH_CHECKED(TYPE, NAME, ...)                    \
+  [&] {                                                                 \
+    switch (TYPE) {                                                     \
+      __VA_ARGS__                                                       \
+      default:                                                          \
+        throw std::invalid_argument(fmt::format(                        \
+            "Unhandled dtype %s for %s", dtype_to_string(TYPE), NAME)); \
+    }                                                                   \
+  }()
+
```

Comment:
Not clear to me why this needs to be wrapped in a lambda? Am I missing something or can we just remove that?

Reply:
The original design was to enable something like `xxx = MLX_SWITCH_TYPES(...)`, I have updated the PR to remove it.

Reply:
I see, thanks for the explanation.